src/dbnode/persist/fs/streaming_write.go

```diff
@@ -0,0 +1,259 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/m3db/bloom/v4"
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/x/checked"
+	"github.com/m3db/m3/src/x/context"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+// StreamingWriter writes int data fileset without intermediate buffering.
+// Writes must be lexicographically ordered by the id.
+type StreamingWriter interface {
+	Open() error
+	Write(ctx context.Context, encoder encoding.Encoder, id ident.ID, encodedTags []byte) error
+	Close() error
+}
+
+// StreamingWriterOptions in the options for the StreamingWriter.
+type StreamingWriterOptions struct {
+	Options             Options
+	NamespaceID         ident.ID
+	ShardID             uint32
+	BlockStart          time.Time
+	BlockSize           time.Duration
+	VolumeIndex         int
+	PlannedRecordsCount uint
+}
+
+type streamingWriter struct {
+	opts         StreamingWriterOptions
+	writer       *writer
+	writerOpts   DataWriterOpenOptions
+	data         []checked.Bytes
+	currIdx      int64
+	prevIDBytes  []byte
+	summaryEvery int64
+	bloomFilter  *bloom.BloomFilter
+	indexOffset  int64
+	summaries    int
+}
+
+func NewStreamingWriter(opts StreamingWriterOptions) (StreamingWriter, error) {
+	w, err := NewWriter(opts.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	writerOpts := DataWriterOpenOptions{
+		BlockSize: opts.BlockSize,
+		Identifier: FileSetFileIdentifier{
+			Namespace:   opts.NamespaceID,
+			Shard:       opts.ShardID,
+			BlockStart:  opts.BlockStart,
+			VolumeIndex: opts.VolumeIndex,
+		},
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	m, k := bloom.EstimateFalsePositiveRate(
+		opts.PlannedRecordsCount,
+		opts.Options.IndexBloomFilterFalsePositivePercent(),
+	)
+	bloomFilter := bloom.NewBloomFilter(m, k)
+
+	summariesApprox := float64(opts.PlannedRecordsCount) * opts.Options.IndexSummariesPercent()
+	summaryEvery := 0
+	if summariesApprox > 0 {
+		summaryEvery = int(math.Floor(float64(opts.PlannedRecordsCount) / summariesApprox))
+	}
+
+	return &streamingWriter{
+		opts:         opts,
+		writer:       w.(*writer),
+		writerOpts:   writerOpts,
+		summaryEvery: int64(summaryEvery),
+		data:         make([]checked.Bytes, 2),
+		bloomFilter:  bloomFilter,
+	}, nil
+}
+
+func (w *streamingWriter) Open() error {
+	if err := w.writer.Open(w.writerOpts); err != nil {
+		return err
+	}
+	w.indexOffset = 0
+	w.summaries = 0
+	w.prevIDBytes = nil
+	return nil
+}
+
+func (w *streamingWriter) Write(
+	ctx context.Context,
+	encoder encoding.Encoder,
```

Comment:
I would just take a `[]byte` here for encoded data so that the caller can control the lifecycle of the data, etc.

Reply:
Would structure the args similar to the other fs.Writer so that Head/Tail can be passed in but also allow a single buffer to be written:
```go
func (w *streamingWriter) Write(
  id ident.ID,
  encodedTags []byte,
  data []byte,
) error {
  w.singleBufferArray[0] = data
  return w.WriteAll(id, encodedTags, w.singleBufferArray)
}

func (w *streamingWriter) WriteAll(
  id ident.ID,
  encodedTags []byte,
  data [][]byte,
) error {
 // .. write multiple data buffers
}

```

Reply:
Got it. I think we need just the `WriteAll` though.